### PR TITLE
Support MegEngine TracedModule format

### PR DESCRIPTION
**Pull request by dingshaohua960303**
Hi, I'm trying to support MegEngine TracedModule (.tm a pickle file) format in netron, and want to merge the new feature to the main branch. [MegEngine](https://github.com/MegEngine/MegEngine) is an open-source deep learning framework. Could you please help review this code, thanks!

the sample models:
[shufflenet_v2_x1_0.tm.zip](https://github.com/lutzroeder/netron/files/9257733/shufflenet_v2_x1_0.tm.zip)
[mobilenetv1.tm.zip](https://github.com/lutzroeder/netron/files/9257882/mobilenetv1.tm.zip)

#607

### CODE REVIEW

File: source/python.js [hunk 1]
```diff
@@ -5574,7 +5574,7 @@ python.Unpickler.BinaryReader = class {
     float64() {
         const position = this._position;
         this.skip(8);
-        return this._view.getFloat64(position, true);
+        return this._view.getFloat64(position, false);
```
Comment (dingshaohua960303) [inline source/python.js:5577]: It seems that pickle save float object in big-endian, https://github.com/python/cpython/blob/a0c7df09689769f3016dc9145eb56341aac631a9/Lib/pickle.py#L784

File: source/hdf5.js [hunk 2]
```diff
@@ -210,7 +210,7 @@ hdf5.Variable = class {
         switch (this._dataLayout.layoutClass) {
             case 1: // Contiguous
                 if (this._dataLayout.address) {
-                    return this._reader.at(this._dataLayout.address).read(this._dataLayout.size);
+                    return this._reader.at(this._dataLayout.address).stream(this._dataLayout.size);
```
Comment (lutzroeder) [inline source/hdf5.js:213]: Please rebase to main.

File: source/index.html [hunk 3]
```diff
@@ -366,11 +366,11 @@
...
+    <input type="file" id="open-file-dialog" class="open-file-dialog" multiple="false" accept=".onnx, .ort, .pb, .meta, .tflite, .lite, .tfl, .keras, .h5, .hd5, .hdf5, .json, .model, .mar, .params, .param, .armnn, .mnn, .ncnn, .tnnproto, .tmfile, .ms, .om, .nn, .uff, .rknn, .xmodel, .kmodel, .paddle, .pdmodel, .pdiparams, .pdparams, .pdopt, .nb, .dnn, .dlc, .cmf, .mlmodel, .mlpackage, .caffemodel, .pbtxt, .prototxt, .pkl, .pt, .pth, .ptl, .t7, .joblib, .cfg, .xml, .zip, .tar ,.tm">
     <!-- Preload fonts to workaround Chrome SVG layout issue -->
     <div style="font-weight: normal; color: rgba(0, 0, 0, 0.01); user-select: none;">.</div>
     <div style="font-weight: bold; color: rgba(0, 0, 0, 0.01); user-select: none;">.</div>
     <div style="font-weight: bold; color: rgba(0, 0, 0, 0.01); user-select: none;">.</div>
 </div>
 </body>
-</html>
\ No newline at end of file
+</html>
```
Comment (lutzroeder) [inline source/index.html:376]: Formatting.

File: source/mge-metadata.json [hunk 4]
```diff
@@ -0,0 +1,2991 @@
...
+    {
+        "name": "megengine.module.batchnorm.BatchNorm2d",
+        "schema": {
+            "category": "Normalization"
+        }
+    },
+    {
+        "name": "megengine.module.concat.Concat",
+        "schema": {
+            "category": "Layer"
```
Comment (lutzroeder) [inline source/mge-metadata.json:2603]: Fix category for multiple operators. For example: `Concat`, `getitem`, `floordiv`...

File: source/mge.js [hunk 5]
```diff
@@ -0,0 +1,1005 @@
...
+                this.np_dtype = np_dtype
+                this.qmin = qmin
+                this.qmax = qmax
+                this.is_signed = is_signed
+            }
+        })
+
+    }
+
+    debug(file) {
```
Comment (lutzroeder) [inline source/mge.js:723]: This file contains parts copied from other sources. Please remove and update.

**Review by lutzroeder — CHANGES_REQUESTED** [review]
- [x] Fix `make lint` issues
- [x] Add test files to `./test/models.json`
- [ ] `node ./test/models.json megengine` is failing

```
~/Projects/netron: node ./test/models.js megengine
megengine/mobilenetv1.tm
Invalid model format 'megengine v1.10.0'.
```

File: source/mge-metadata.json [hunk 6]
```diff
@@ -0,0 +1,2875 @@
+
```
Comment (lutzroeder) [inline source/mge-metadata.json:1]: Rename file to `megengine-metadata.json`. Format using 2 leading space.

File: source/mge.js [hunk 7]
```diff
@@ -0,0 +1,988 @@
+
+// Experimental
+
+var megengine = megengine || {};
+var python = python || require('./python');
+var base = base || require('./base');
+
+megengine.ModelFactory = class {
```
Comment (lutzroeder) [inline source/mge.js:8]: Rename file to `megengine.js` to match namespace.

File: source/mge.js [hunk 8]
```diff
@@ -0,0 +1,988 @@
...
+
+    match(context) {
+        return megengine.Container.open(context);
+    }
+
+    open(context, match) {
+        const identifier = context.identifier;
+        return megengine.Metadata.open(context).then((metadata) => {
```
Comment (lutzroeder) [inline source/mge.js:16]: Replace with default implementation? Why is a custom `Metadata` implementation needed?

`return context.metadata('megengine-metadata.json').then((metadata) => {`

File: source/mge.js [hunk 9]
```diff
@@ -0,0 +1,988 @@
...
+                this.qmin = qmin;
+                this.qmax = qmax;
+                this.is_signed = is_signed;
+            }
+        });
+
+    }
+};
+
+megengine.Container = class {
```
Comment (lutzroeder) [inline source/mge.js:725]: Merge `megengine.Container` and `megengine.Container.Pickle` into a single `megengine.Pickle` class.

File: source/mge.js [hunk 10]
```diff
@@ -0,0 +1,988 @@
...
+    }
+
+    get format() {
+        this._unpickle();
+        var version = "1.6";
+        var dump_info = this._graphs[0].data.dump_info;
+        if (dump_info !== undefined) {
+            version = dump_info.version;
+        }
+        return 'megengine v' + version;
```
Comment (lutzroeder) [inline source/mge.js:765]: `MegEngine` to use same casing as repo and Pip package?

File: source/mge.js [hunk 11]
```diff
@@ -0,0 +1,988 @@
...
+        const execution = new megengine.Execution(null, this._exceptionCallback);
+        const unpickler = python.Unpickler.open(data, execution);
+
+        this._stream = null;
+        this._exceptionCallback = null;
+
+        const module_source_map = new Map();
+        const deserialized_objects = new Map();
+        unpickler.persistent_load = (saved_id) => {
+            const typename = saved_id.shift();
```
Comment (lutzroeder) [inline source/mge.js:793]: Copied code that is not being used? Please review and remove unused code.

File: source/mge.js [hunk 12]
```diff
@@ -0,0 +1,988 @@
...
+                    }
+                    return deserialized_objects.get(root_key);
+                }
+                default: {
+                    throw new megengine.Error("Unsupported persistent load type '" + typename + "'.");
+                }
+            }
+        };
+        const obj = unpickler.load();
+        if (!obj) {
```
Comment (lutzroeder) [inline source/mge.js:831]: Inline `find` and `findModule` method, simplify.

**Review by lutzroeder — CHANGES_REQUESTED** [review]

**Review by lutzroeder — CHANGES_REQUESTED** [review]
- [ ] `node ./test/models.json megengine` is failing

```
~/Projects/netron: node ./test/models.js megengine
megengine/mobilenetv1.tm
Invalid model format 'megengine v1.10.0'.
```

File: source/mge.js [hunk 13]
```diff
@@ -0,0 +1,988 @@
...
+        super(sources, exceptionCallback);
+        this.registerType('megengine.tensor.Parameter', class {
+            constructor(data, dtype, device) {
+                this.data = data;
+                this.dtype = dtype;
+                this.device = device;
+            }
+        });
+        this.registerType('megengine.traced_module.pytree.TreeDef', class {
+            toString() {
```
Comment (lutzroeder) [inline source/mge.js:647]: These methods do not exist in the Python implementation. Formatting should be done in client code, `registerType` should only be used to match Python execution and serialization.

**Review by lutzroeder — CHANGES_REQUESTED** [review]

File: source/mge.js (hunk 13 above)
Comment (dingshaohua960303) [reply source/mge.js]: `toString` method is needed in serialization, the treedef object may be a key of `argdef_graph_map`, without it the dict will be loaded as `[[object]] : xxxx`.

`formatString` method has been moved to `megengine.Utility` as a static method.

File: source/mge.js (hunk 10 above)
Comment (dingshaohua960303) [reply source/mge.js]: the name of Pip package is `MegEngine`,  and I modified the format to `MegEngine + version`

File: source/mge.js (hunk 9 above)
Comment (dingshaohua960303) [reply source/mge.js]: done

File: source/mge.js (hunk 8 above)
Comment (dingshaohua960303) [reply source/mge.js]: done

File: source/mge.js (hunk 7 above)
Comment (dingshaohua960303) [reply source/mge.js]: done

File: source/mge-metadata.json (hunk 6 above)
Comment (dingshaohua960303) [reply source/mge-metadata.json]: done

File: source/mge.js (hunk 11 above)
Comment (dingshaohua960303) [reply source/mge.js]: done

File: source/mge.js (hunk 12 above)
Comment (dingshaohua960303) [reply source/mge.js]: done

**Pull request closed without merging**